api/storage_service.cc

```diff
@@ -1367,8 +1367,18 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
     });
 
     ss::get_effective_ownership.set(r, [&ctx, &ss] (std::unique_ptr<http::request> req) {
-        auto keyspace_name = req->param["keyspace"] == "null" ? "" : validate_keyspace(ctx, req->param);
-        return ss.local().effective_ownership(keyspace_name).then([] (auto&& ownership) {
+        auto keyspace = req->param["keyspace"] == "null" ? "" : validate_keyspace(ctx, req->param);
```

Comment:
Hi @avikivity, I have a question just to ensure that I understand this line of code correctly.

Is the following path some **special** kind of ownership query?
`/storage_service/ownership/null`

I wonder, because at glance I thought that such scenario would be executed when keyspace parameter is not specified - that was the first thing that came to my mind when I saw `null`.

However, it seems, that in case of `/storage_service/ownership/` a completely different path is taken.

According to api/api-doc/storage_service.json:
```
{
         "path":"/storage_service/ownership/",
         "operations":[
            {
               "method":"GET",
               "summary":"The mapping from token -> % of cluster owned by that token",
               "type":"array",
               "items":{
                  "type":"mapper"
               },
               "nickname":"get_ownership",
               "produces":[
                  "application/json"
               ],
               "parameters":[
               ]
            }
         ]
      }
```
and
```
"path":"/storage_service/ownership/{keyspace}",
         "operations":[
            {
               "method":"GET",
               "summary":"Effective ownership is % of the data each node owns given the keyspace",
               "type":"array",
               "items":{
                  "type":"mapper"
               },
               "nickname":"get_effective_ownership",
               "produces":[
                  "application/json"
               ],
               "parameters":[
                  {
                     "name":"keyspace",
                     "description":"The keyspace to fetch information about",
                     "required":true,
                     "allowMultiple":false,
                     "type":"string",
                     "paramType":"path"
                  },
                  {
                     "name":"cf",
                     "description":"Column family name",
                     "required":false,
                     "allowMultiple":false,
                     "type":"string",
                     "paramType":"query"
                  }
               ]
            }
         ]
```

So, as far as I understand in case of `/storage_service/ownership` ss::get_ownership is used, whether in case of `/storage_service/ownership/keyspace` ss::get_effective_ownership is invoked. 


Perhaps, if `/storage_service/ownership/null` is a special query, then additional tests need to be added to this PR.


Also, in spite of adding tests that execute `/storage_service/ownership` it seems, that they are not valid when tablets are enabled. `ss::get_ownership` is called and it invokes `storage_service::get_ownership()` that uses only `sorted_tokens()` - so, as far as I understand it does not support tablets. Do I understand it correctly?

Perhaps, this PR needs to be enhanced by taking into account also the other handler.

Reply:
> Is the following path some special kind of ownership query?
> /storage_service/ownership/null

Yes. This is used by `nodetool status`, when invoked without a keyspace argument.

Reply:
The difference between `ss::ownersip()` and `ss::effective_ownership()` is that the latter takes into account how data is replicated in datacenters. The former doesn't.

For tablets, we will need a completely different code path for both: this implementation would use `system.tablets` and not the sorted tokens, as the basis of calculating ownership. Further thinking about it, I think with tablets, we can only provide meaningful implementation for `ss::effective_ownership()`. `ss:ownership()` has even less meaning with tablets than it has with vnodes (and it is already pretty useless with vnodes).

It looks like we will need some changes to `nodetool status` to adjust for tablets. I'm in the process of re-implementing it in C++ currently (we are only doing tablet-related changes in the native nodetool, not the legacy one).

Reply:
Hi @denesb - I introduced a new if statement in handler of `storage_service/ownership` that checks if any of keyspaces uses tablets - in such case BAD_REQUEST is returned. To verify the new logic an additional test case has been introduced.

Is it okay? Could you take a look at it?

Reply:
Would it make sense to push the functionality into a vitual method in effective_replication_map?

Reply:
> Would it make sense to push the functionality into a vitual method in effective_replication_map?

Hi @bhalevy - could you specify which functionality? Are you referring to `storage_service/ownership/null`?

Reply:
> > Would it make sense to push the functionality into a vitual method in effective_replication_map?
> 
> Hi @bhalevy - could you specify which functionality? Are you referring to `storage_service/ownership/null`?

no, I meant the following functionality:

> For tablets, we will need a completely different code path for both: this implementation would use system.tablets and not the sorted tokens, as the basis of calculating ownership.

Reply:
Anyhow, making get_ranges virtual seem to go in exactly that direction.